GCTC-NTGC/gc-digital-talent
# Feature - Add rate limits for api, graphql

Resolves #4185.

h/t to @esizer.

## Notes
Since we only have one "api" endpoint right now in `api/support/tickets, 100 requests per minute should not be a problem. However, since almost every page for **admin** and **talentsearch** run one or more GraphQL queries, 100 requests per minute for "graphql" might be too low a number.

- **Update 1**: default request limit set to 600.
- **Update 2**: added a throttle on the `createUser` GraphQL mutation of 100 requests per minute for future reference (and, since no one should be creating more than one user per second!)

## Screenshots
![Screen Shot 2022-12-08 at 14 32 44](https://user-images.githubusercontent.com/3046459/206563296-fd7f58c6-8523-46a7-b9bf-00c169e367c8.png)
![Screen Shot 2022-12-08 at 14 34 06](https://user-images.githubusercontent.com/3046459/206563302-24432f6e-e2da-4fad-aa0a-a43d2f4f192d.png)
![Screen Shot 2022-12-08 at 14 37 16](https://user-images.githubusercontent.com/3046459/206563303-15c5caa4-dd1c-4f4f-885c-95a6ed51f5df.png)
![Screen Shot 2022-12-08 at 15 27 09](https://user-images.githubusercontent.com/3046459/206563307-c38e8d6d-45ed-4459-bd76-093c027c8b86.png)


## Steps to test

1. Compile common `npm run intl-compile --workspace=common`
2. Build workspaces `npm run production --workspaces --if-present`
3. Add [`APP_RATE_LIMIT=10`] in (api/.env)
4. Change GraphQL me query in schema.graphql from `me: User @auth` to `me: User @auth @throttle(maxAttempts: 1)`
5. Refresh API (refresh_api.sh)
6. Navigate to /support
7. Verify singular toast appears in English with correct text
8. Switch to French
9. Verify singular toast appears in French with correct text
10. Navigate to /search
11. Verify singular toast appears in English with correct text
12. Switch to French
13. Verify singular toast appears in French with correct text
14. Navigate to /admin/dashboard
15. Verify singular toast appears in English with correct text
16. Switch to French
17. Verify singular toast appears in French with correct text
18. Navigate to /login
19. Verify Too Many Requests text
